File: azkar/apple/macOS/Window/CardsPage.swift
import AppKit
import SwiftUI

/// The shortcut, how general azkar are picked, and how cards look.
struct CardsPage: View {
  @ObservedObject var model: SettingsModel

  var body: some View {
    Form {
      Section {
        Row(title: "Count / close shortcut", symbol: "keyboard", color: .blue) {
          ShortcutRecorder(model: model)
        }
        if let error = model.status.hotkeyError {
          Label(error, systemImage: "exclamationmark.triangle.fill").foregroundStyle(.orange)
        }
      } footer: {
        Footnote(
          "Each press — or a click on the card — counts once, and the last one closes it. × closes a card straight away."
        )
      }

      Section("General azkar") {
        SwiftUI.Picker(selection: $model.config.order) {  // Core has its own `Picker`
          Text("Random").tag(Order.random)
          Text("In order").tag(Order.sequential)
        } label: {
          Row(title: "Order", symbol: "shuffle", color: .teal) { EmptyView() }
        }
        Toggle(isOn: $model.config.repeatGeneral) {
          Row(
            title: "Repeat counts", symbol: "repeat", color: .green,
            detail: "Off: one tap each, even for ×100 azkar."
          ) { EmptyView() }
        }
        .toggleStyle(.switch)
      }

      Section("Cards") {
        Stepper(value: $model.config.maxStack, in: 1...10) {
          Row(title: "Cards on screen at most", symbol: "rectangle.stack.fill", color: .pink) {
            Text("\(model.config.maxStack)").monospacedDigit().foregroundStyle(.secondary)
          }
        }
        Row(title: "Text size", symbol: "textformat.size", color: .gray) {
          // No `step:` (it draws tick marks); whole points instead.
          Slider(
            value: Binding(get: { model.config.fontSize }, set: { model.config.fontSize = $0.rounded() }),
            in: 14...36
          )
          .frame(width: 200)
          Text("\(Int(model.config.fontSize))").monospacedDigit().foregroundStyle(.secondary).frame(width: 24)
        }
        CardPreview(fontSize: model.config.fontSize, hint: model.config.hotkey.display)
      }
    }
    .formStyle(.grouped)
    .navigationTitle("Cards")
  }
}

/// Roughly what a card looks like at the chosen text size.
private struct CardPreview: View {
  let fontSize: Double
  let hint: String

  var body: some View {
    VStack(spacing: 10) {
      HStack(spacing: 8) {
        Image(systemName: "xmark.circle.fill").foregroundStyle(.tertiary)
        Text("×3").font(.caption.bold()).foregroundStyle(.orange)
        Spacer()
        Text("أذكار الصباح  ·  7 من 25").font(.caption.weight(.semibold)).foregroundStyle(.orange)
      }
      Text("سُبْحَانَ اللَّهِ وَبِحَمْدِهِ، عَدَدَ خَلْقِهِ، وَرِضَا نَفْسِهِ، وَزِنَةَ عَرْشِهِ، وَمِدَادَ كَلِمَاتِهِ")
        .font(.system(size: fontSize))
        .lineSpacing(fontSize * 0.3)
        .multilineTextAlignment(.trailing)
        .frame(maxWidth: .infinity, alignment: .trailing)
      Text("انقر أو \(hint)").font(.caption2).foregroundStyle(.tertiary)
        .frame(maxWidth: .infinity, alignment: .leading)
    }
    .padding(18)
    .frame(maxWidth: 440)
    .background(Color(white: 0.13), in: RoundedRectangle(cornerRadius: 14, style: .continuous))
    .environment(\.colorScheme, .dark)
    .frame(maxWidth: .infinity)
    .padding(.vertical, 6)
  }
}

File: azkar/kotlin/shared/src/commonMain/kotlin/com/wildduck/azkar/ui/CardsPage.kt
// The shortcut, how general azkar are picked, and how cards look.
package com.wildduck.azkar.ui

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.layout.widthIn
import androidx.compose.material3.FilterChip
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Slider
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import com.wildduck.azkar.app.CardOnScreen
import com.wildduck.azkar.app.Features
import com.wildduck.azkar.app.ProblemAction
import com.wildduck.azkar.app.ProblemKey
import com.wildduck.azkar.app.ShortcutMode
import com.wildduck.azkar.app.UiState
import com.wildduck.azkar.core.Card
import com.wildduck.azkar.core.Order
import com.wildduck.azkar.core.Session
import com.wildduck.azkar.core.Zikr
import kotlin.math.roundToInt

private const val PREVIEW_TEXT =
    "سُبْحَانَ اللَّهِ وَبِحَمْدِهِ، عَدَدَ خَلْقِهِ، وَرِضَا نَفْسِهِ، وَزِنَةَ عَرْشِهِ، وَمِدَادَ كَلِمَاتِهِ"

@Composable
fun CardsPage(ui: UiState, features: Features, actions: AzkarActions) {
    val config = ui.config
    Page("Cards") {
        if (features.shortcut) {
            Section(
                footer = "Each press — or a click on the card — counts once, and the last one closes it. " +
                    "× closes a card straight away.",
            ) {
                SettingsRow("Count / close shortcut", Accents.blue) {
                    when (val mode = ui.shortcut.mode) {
                        is ShortcutMode.App -> ShortcutRecorder(
                            hotkey = config.hotkey,
                            style = features.keyStyle,
                            onRecord = { actions.setConfig(config.copy(hotkey = it)) },
                        )
                        is ShortcutMode.SystemSettings -> Row(
                            verticalAlignment = Alignment.CenterVertically,
                            horizontalArrangement = Arrangement.spacedBy(8.dp),
                        ) {
                            Text(ui.shortcut.display.ifEmpty { config.hotkey.display(features.keyStyle) })
                            TextButton(onClick = { actions.perform(ProblemAction.OpenShortcutSettings) }) {
                                Text(mode.hint)
                            }
                        }
                        ShortcutMode.None -> Text(
                            "Not available here",
                            color = MaterialTheme.colorScheme.onSurfaceVariant,
                        )
                    }
                }
                ui.problems.firstOrNull { it.key == ProblemKey.Shortcut }?.let { problem ->
                    Text(
                        problem.text,
                        Modifier.fillMaxWidth().padding(start = 48.dp, end = 14.dp, bottom = 10.dp),
                        color = MaterialTheme.colorScheme.error,
                        style = MaterialTheme.typography.bodySmall,
                    )
                }
            }
        }

        Section("General azkar") {
            SettingsRow("Order", Accents.teal) {
                Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                    FilterChip(
                        selected = config.order == Order.Random,
                        onClick = { actions.setConfig(config.copy(order = Order.Random)) },
                        label = { Text("Random") },
                    )
                    FilterChip(
                        selected = config.order == Order.Sequential,
                        onClick = { actions.setConfig(config.copy(order = Order.Sequential)) },
                        label = { Text("In order") },
                    )
                }
            }
            RowDivider()
            SwitchRow(
                title = "Repeat counts",
                color = Accents.green,
                detail = "Off: one tap each, even for ×100 azkar.",
                checked = config.repeatGeneral,
                onChange = { actions.setConfig(config.copy(repeatGeneral = it)) },
            )
        }

        if (features.cards) {
            Section("Cards") {
                StepperRow(
                    title = "Cards on screen at most",
                    color = Accents.pink,
                    value = config.maxStack.toString(),
                    range = 1..10,
                    current = config.maxStack,
                    onChange = { actions.setConfig(config.copy(maxStack = it)) },
                )
                RowDivider()
                SettingsRow("Text size", Accents.grey) {
                    Slider(
                        value = config.fontSize.toFloat(),
                        onValueChange = {
                            actions.setConfig(config.copy(fontSize = it.roundToInt().toDouble()))
                        },
                        valueRange = 14f..36f,
                        modifier = Modifier.width(180.dp),
                    )
                    Text(
                        config.fontSize.roundToInt().toString(),
                        Modifier.width(26.dp),
                        color = MaterialTheme.colorScheme.onSurfaceVariant,
                        textAlign = TextAlign.End,
                    )
                }
                Box(Modifier.fillMaxWidth().padding(14.dp), contentAlignment = Alignment.Center) {
                    CardView(
                        card = preview,
                        fontSize = config.fontSize,
                        hint = config.hotkey.display(features.keyStyle),
                        onTap = {},
                        onClose = {},
                        modifier = Modifier.widthIn(max = cardWidth),
                    )
                }
            }
        }
    }
}

/** Roughly what a card looks like at the chosen text size. */
private val preview =
    CardOnScreen(0, Card(Zikr(PREVIEW_TEXT, 3), Session.Sabah, 7, 25), badge = "×3", fraction = 0.0)
